pk2m-49 upload config files to gcs bucket

AlexBMes opened the pull request:
New functionality for uploading config directories to cloud

AlexBMes committed:
pk2m-49 upload config files to gcs bucket

AlexBMes committed:
remove commented functions

gbmeuk commented on concierge_scheduler/concierge_cloud.py:
```diff
@@ -0,0 +1,90 @@
+        self.command_mapping[action]()
+
+    # logging
+    def get_logger(self, name):
```
Don't put this into your interface. The interface should be kept clean for its purpose, and in this case it isn't logging.

AlexBMes replied:
Moved out of Interface

gbmeuk commented on concierge_scheduler/concierge_cloud.py:
```diff
@@ -0,0 +1,90 @@
+    def authenticate(self):
+        """Load in the credential file for cloud backup"""
+        f = open(self.credential_file_path, 'r')
+        self._info('Authenticating with {} credentials...', self.credential_file_path)
```
This should be in the implementation

AlexBMes replied:
Had originally put this in there because this part should happen for every implementation, but moved it to `_load_credentials_file()` method.

gbmeuk commented on concierge_scheduler/concierge_cloud.py:
```diff
@@ -0,0 +1,90 @@
+        self._info('Authenticating with {} credentials...', self.credential_file_path)
+        try:
+            self.credential_file = json.load(f)
+        except:
```
Be specific about what exception you're catching otherwise you'll catch everything and could be confusing when you hit a different exception later in the code.

AlexBMes replied:
Changed to catch `json.JSONDecodeError`, and is now only attempts to decode JSON and yaml files, otherwise will just read

gbmeuk commented on concierge_scheduler/concierge_cloud.py:
```diff
@@ -0,0 +1,90 @@
+        except:
+            self._warn('File {} not in JSON format, JSON format required for GCP authentication', self.credential_file_path)
+            self.credential_file = f.read()
+        f.close()
```
Use 
```
with open(self.credential_file_path, 'r') as f:
  self.credential_file = json.load(f)
``` 

and you won't have to close it

AlexBMes replied:
Done

gbmeuk commented on concierge_scheduler/concierge_cloud.py:
```diff
@@ -0,0 +1,90 @@
+
+
+class CloudBackupInterface:
+    def __init__(self, bucket_name, credential_file, config_dir, bucket_folder=''):
```
These signature feels like an implementation detail. Do you need all these passed for every implementation? Can some of them just be arguments to methods?

AlexBMes replied:
The current usage of this class in `concierge_scheduler.py`: 
```python
    ...
    event_admin(zbx_client, cmd_args.config_dir, cmd_args.force_templates).run(cmd_args.command)
elif cmd_args.command in ['upload']: 
    cloud_admin(cmd_args.bucket_name, credential_files[cmd_args.cloud_engine], cmd_args.config_dir, 
       cmd_args.bucket_folder).run(cmd_args.command) 
```
Where `cloud_admin` is the class `GCSAdmin` (based on user input) and the run command runs `upload_local_dir()`.

I feel that all the arguments would be required for all cloud implementations:
* `bucket_name` - As far as I can see only Azure uses `blob` instead of `bucket`, but this variable would be required for all cloud storage functionality
* `credential_file` - This could be changed to `storage_client` which would be the authenticated client object, but since authentication happens within the same class as the other methods, it seems worth authenticating on class init. 
* `config_dir` - Path to directory where the configuration files for `upload` are locate. This would also apply for a download if that was implemented, as the directory to save the files to.
* `bucket_folder` - This is optional, so moved to only `GCSAdmin` class, but possibly still a useful feature across different cloud providers, as this is the folder within the bucket that the `config_dir` will be uploaded to. So files are `<bucket>/<bucket_folder>/<config_dir name>/<file>`

gbmeuk commented on concierge_scheduler/concierge_cloud.py:
```diff
@@ -0,0 +1,90 @@
+        self.__LOG.log(logging.ERROR, message.format(*args))
+        sys.exit(-1)
+
+    def authenticate(self):
```
This doesn't feel like an authentication method. Seems like it is a load-a-file method. We will want to authenticate so this should be part of the interface but most likely implementation will be different, so maybe you just want to `raise NotImplementedError`

AlexBMes replied:
See above, moved  code to `_load_credential_file()`

gbmeuk commented on concierge_scheduler/concierge_cloud.py:
```diff
@@ -0,0 +1,90 @@
+        self.authenticate()
+
+    def authenticate(self):
+        super().authenticate()
```
you don't need to `super()` each method

AlexBMes replied:
Now only use `super()` in `__init__()`

gbmeuk commented on concierge_scheduler/concierge_cloud.py:
```diff
@@ -0,0 +1,90 @@
+                self.config_files.append(os.path.join(root, file))
+
+
+class GCSAdmin(CloudBackupInterface):
```
separate the implementation and interface classes into separate modules

AlexBMes replied:
Interface is in `concierge_cloud.py`, `GCSAdmin` implementation is in `concierge_gcs.py`.

gbmeuk commented on concierge_scheduler/concierge_cloud.py:
```diff
@@ -0,0 +1,90 @@
+            self.credential_file = f.read()
+        f.close()
+
+    def upload_local_dir(self):
```
I think `upload_local_dir()` needs to be simply `raise NotImplementedError` and if there is a method which collects up local files and directories which is consistent across all implementations, this should be implemented in the interface

AlexBMes replied:
This now just `raise NotImplementedError` and the method for getting the files is `_get_config_files()` in the interface.

gbmeuk reviewed (changes requested):
Some changes here but good work still.

AlexBMes committed:
pk2m-49 Separate modules and refactor based on comments

AlexBMes committed:
pk2m-49 Move init arguments to method

Commit:
implement formal interface

AlexBMes committed:
pk2m-49 Move logging, implement interface

AlexBMes committed:
Merge pull request #11 from mesoform/feat/pk2m-49/interface

implement formal interface, move logging out of classes

gbmeuk commented on concierge_scheduler/concierge_gcs.py:
```diff
@@ -1,69 +1,52 @@
+            else:
+                raise KeyError('Key "project_id" not found in credential file. Invalid credential File')
+
+    def set_storage_location(self, location: str):
```
A more Pythonic way of doing this would be

```
def __init__(self, credential_file_path, config_dir):
    self.__storage_location = None
```
```
    @property
    def storage_location(self):
        return self.__storage_location
        
    @storage_location.setter
    def storage_location(self, location: str):
        self._storage_location = location
```

gbmeuk replied:
Ah, `__storage_location` already exists in the base class

gbmeuk replied:
probably even better as a
```
@abstractmethod
@property
...
```
in the interface

AlexBMes replied:
Have updated. Did not use `@abstractmethod` for the property, as kept encountering errors attempting to implement that way. Instead did:
```python 
@property
def storage_location(self) -> str:
    return self._storage_location

@storage_location.setter
def storage_location(self, location):
    self._storage_location = location
```

gbmeuk replied:
Yes. True. They're already concrete methods there.

gbmeuk commented on concierge_scheduler/concierge_gcs.py:
```diff
@@ -1,69 +1,52 @@
+            if 'project_id' in credential_file:
+                storage_credentials = service_account.Credentials.from_service_account_info(credential_file)
+                client = storage.Client(project=credential_file['project_id'], credentials=storage_credentials)
+                return client
```
trivial point, you could just

```
return storage.Client(project=credential_file['project_id'], credentials=storage_credentials)
```

AlexBMes replied:
done

gbmeuk commented on concierge_scheduler/concierge_gcs.py:
```diff
@@ -1,69 +1,52 @@
+        self._client = self.authenticate()
+        self.config_dir = config_dir
 
     def authenticate(self):
```
Doc string, type hint

gbmeuk replied:
a type hint here prevents your IDE from throwing a warning about unresolved reference on `bucket = self._client.get_bucket(self._storage_location)`

AlexBMes replied:
Updated return type

gbmeuk commented on concierge_scheduler/concierge_gcs.py:
```diff
@@ -1,69 +1,52 @@
+                config_files.add(os.path.join(root, file))
+        return config_files
+
+    def upload(self, upload_list, folder=''):
```
type hints

AlexBMes replied:
Updated types

gbmeuk commented:
Some minor improvements for maintainability but generally looks good

AlexBMes committed:
pk2m-49 Changes from comments, update naming and docs

AlexBMes committed:
Merge branch 'feat/pk2m-49/interface' into feat/pk2m-49/GCS-uploads

Commit:
private attribute on base class

```
gcs_backup = GCSBackup("/Users/gaz/.config/build-services-dev.json", "/another/path", "bucket_name")
dir(gcs_backup)
['__abstractmethods__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_client', '_storage_location', 'assemble_upload_list', 'authenticate', 'config_dir', 'credential', 'storage_location', 'upload']
gcs_backup._storage_location = "test"
gcs_backup.storage_location
'test'
```
Vs
```
gcs_backup = GCSBackup("/Users/gaz/.config/build-services-dev.json", "/another/path", "bucket_name")
dir(gcs_backup)
['_CloudBackupInterface__storage_location', '__abstractmethods__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_client', 'assemble_upload_list', 'authenticate', 'config_dir', 'credential', 'storage_location', 'upload']
gcs_backup.__storage_location = "test"
gcs_backup.storage_location
'bucket_name'
```
